### PR TITLE
Resolve mismatch of tree and node from CFG and AST when handling postfix operation

### PR DESCRIPTION
**Pull request by AndrewShf**
Screen Shot 2022-07-18 at 5 36 51 PM

Corresponding code is
```
        int d = 1;
        int i = 1;
        d *= i--;
```

For node like `d * tempPostfix#num0`, the cfg side will try to query the value from factory, and the corresponding tree is `d * i--`. The factory recursively query the value of `i--`, and recursively to value of `i`. When the factory analyzes tree `i`, the last step is to apply inferredValue from the cfg side, which again goes back to cfg, and turns `i` to node `i`, which isn't the operands of `d * tempPostfix#num0` and thus returns null in `AbstractAnalysis.getValue(Node n)`. After the analyzing of `i` is complete, program goes one level up analyzing `i--` again, and the last step is to apply inferredValue from the cfg side, which again goes back to cfg, and turns `i--` to node `tempPostfix#num0 = i`, which returns null because of the same reason as before. 

The reasoning of my change is, tree `i--` corresponds to node `tempPostfix#num0 = i`, and when we query the value of `i--`, we actually want to know the value of `i` before decreasing, which is `tempPostfix#num0` in the assignment. But we cannot use the assignment target `tempPostfix#num0` directly, as the variable is under initialization. We can use the expression of the assignment, which is `i`.

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java [hunk 1]
```diff
@@ -200,6 +200,14 @@ public AnalysisResult<V, S> getResult() {
             assert !n.isLValue() : "Did not expect an lvalue, but got " + n;
             if (!currentNode.getOperands().contains(n)
                     && !currentNode.getTransitiveOperands().contains(n)) {
+                if (n instanceof AssignmentNode) {
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java:203]: Please add some comments that explain this block.
Should it be special-cased to only apply to assignments from post-fix operations?

File: dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java [hunk 2]
```diff
@@ -200,6 +200,21 @@ public AnalysisResult<V, S> getResult() {
             assert !n.isLValue() : "Did not expect an lvalue, but got " + n;
             if (!currentNode.getOperands().contains(n)
                     && !currentNode.getTransitiveOperands().contains(n)) {
+                // if n.getTree.getKind is a postfix tree, we check if the target of the assignment
+                // is contained in the operands or not.
+                if (n instanceof AssignmentNode) {
+                    Tree.Kind nKind = n.getTree().getKind();
```
Comment (AndrewShf) [inline dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java:206]: the `getKind()` method makes currentNode nullable again (gives `deference.of.nullable` when `currentNode.getOperands()`), and the method is in java library.

Comment (AndrewShf) [reply dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java]: already make a pr to eisop jdk

File: dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java [hunk 3]
```diff
@@ -200,6 +200,23 @@ public AnalysisResult<V, S> getResult() {
...
+                // temporary variable for CFG but not for the AST side,
+                // thus cause inconsistency.
+                // See https://github.com/eisop/checker-framework/pull/304
+                if (n instanceof AssignmentNode) {
+                    Tree.Kind nKind = ((AssignmentNode) n).getTree().getKind();
+                    if (nKind != Tree.Kind.POSTFIX_DECREMENT
+                            && nKind != Tree.Kind.POSTFIX_INCREMENT) {
+                        return null;
+                    }
+                    Node expr = ((AssignmentNode) n).getExpression();
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java:213]: Can you add a one-sentence summary of what this does?
So it's clear here that there is some mis-match for post-fix ops, but not clear what is done about it.

**Review by wmdietl — COMMENTED** [review]
Is there a test you can add for this?

File: dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java (hunk 3 above)
Comment (AndrewShf) [reply dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java:213]: be more specific about it.

**Comment by AndrewShf** [comment]
> Is there a test you can add for this?

think about why IntVal passed, and check valuetransfer

**Comment by AndrewShf** [comment]
> Is there a test you can add for this?

For now, it's a bit hard to create a test case that can fail existing typecheckers in CF. I looked `Signedness Checker, UnitsChecker` they all do something similar to replace the type annotation in `visitBinary` in `xxTreeAnnotator`.

**Comment by AndrewShf** [comment]
try simplifying units checker

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -639,13 +639,6 @@ public Void visitIdentifier(IdentifierTree node, AnnotatedTypeMirror type) {
             return null;
         }
 
-        // The result of a binary operation is always non-null.
-        @Override
-        public Void visitBinary(BinaryTree node, AnnotatedTypeMirror type) {
```
Comment (AndrewShf) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessAnnotatedTypeFactory.java:644]: keep this for performance reason; investigate the java8 problem.

**Comment by AndrewShf** [comment]
This PR is supposed to fix the hacky code in `NullnessAnnotatedTypeFactory.java.visitBinary(BinaryTree node, AnnotatedTypeMirror type)`. But the way I fix it is still hacky by handling a special case in `AbstractAnalysis.java` In the end, I found out there was no test case we could add to CF (as I found this problem in PICO), and this fix only solved only one particular case, we decided to use the existing solution.

**Pull request closed without merging**